### PR TITLE
DNA improvements

### LINKED ISSUE #221: Implement improvements on DNA spec
There are some changes to the DNA structure that we should implement:

- [x] Switching from arrays to objects for zomes and capabilities, with the (zome/cap) name being the property name within the encompassing object (the way hcdev currently creates JSON from a folder tree)
- [x] Getting rid of "signature" in function definitions as decided here: https://github.com/holochain/holochain-rust/issues/134
- [x] Renaming "fn_declarations" to "functions"
- [ ] Adding links ("links_to" and "linked_from") as specified in https://github.com/holochain/app-spec-rust/pull/1

Not sure if while at it we would also do:
- [ ] Adding signals as specified in https://github.com/holochain/holochain-rust/blob/develop/doc/architecture/decisions/0013-signals-listeners-model-and-api.md
..since the ADR is not accepted yet..

The main intention behind this ticket is to get the spec app to be at least parsable by Holochain.

### PR DESCRIPTION
**Pull request by lucksus**
closes #221 dna improvements

# Change DNA spec to match app spec and hcdev

- [x] "fn_declarations" -> "functions"
- [x] get rid of "signature" (#134)
- [x] change arrays to maps for zomes, capabilites and entry_types and extract name from inner struct to mapping (matching with https://github.com/holochain/holochain-cmd)
- [x] add "linked_from" as used in the app spec (https://github.com/holochain/app-spec-rust/pull/1/files#diff-4f4d6fab1698c4be58c54cb3c2e563aeR10)

### CODE REVIEW

**Comment by lucksus** [comment]
A HoloSqape build with this branch of holochain-rust can read the DNA in the app-spec-rust dist directory: https://github.com/holochain/app-spec-rust/pull/1/files#diff-c30b6726492851bac2ff0ab8ea3fb01d
screen shot 2018-08-28 at 00 14 24

That file is almost purely the output of `hcdev package --strip-meta` with manually added `"code": { .. }` around WASM.
@sphinxc0re

**Comment by lucksus** [comment]
@zippy, @sphinxc0re has added such meta fields to the output produced by `hcdev`, which I had to avoid with `--strip-meta`.. but yeah, maybe we want to include that in the DNA spec.

File: core_api/src/lib.rs [hunk 1]
```diff
@@ -100,7 +100,11 @@ impl Holochain {
             }
         });
 
-        match receiver.recv_timeout(Duration::from_millis(1000)) {
+        // TODO: what is the right timeout?
```
Comment (thedavidmeister) [inline core_api/src/lib.rs]: haha "the right timeout"

File: dna/src/lib.rs [hunk 2]
```diff
@@ -373,14 +363,14 @@ pub mod tests {
     fn parse_with_defaults_zome() {
         let dna = Dna::new_from_json(
             r#"{
-                "zomes": [
-                    {}
-                ]
+                "zomes": {
+                    "0": {}
```
Comment (thedavidmeister) [inline dna/src/lib.rs]: i won't reject the PR for this, but it is confusing to have strings that look like indexes

**Pull request merged**